### DISCUSSION TITLE
交叉编译设置后不生效

### DISCUSSION THREAD

**Discussion by liangzc1** [xmake-io/xmake#5681] [Q&A]
toolchain("T41")
    set_kind("standalone")
    set_toolset("cc", "mips-linux-gnu-gcc")
    set_toolset("cxx", "mips-linux-gnu-g++")
    set_toolset("ld", "mips-linux-gnu-ld")
    set_toolset("ar", "mips-linux-gnu-ar")
    set_toolset("ex", "mips-linux-gnu-execstack")
    set_toolset("strip", "mips-linux-gnu-strip")
    set_toolset("as", "mips-linux-gnu-as")
toolchain_end()
-- 设置编译器
set_languages("c++11")

-- 定义目标 sample-Encoder-video
target("sample-Encoder-video")
    set_toolchains("T41")
    set_kind("binary")
    add_links("pthread", "m", "rt")
    -- add_packages("T41")

    add_includedirs("T41/libimp-samples/")
    add_files("T41/libimp-samples/sample-common.c", "T41/libimp-samples/sample-Encoder-video.c")

    -- 在编译前打印工具链的 cc
    before_build(function (target)
        local toolchain = target:toolchain("cc")
        if toolchain then
            local cc = toolchain:tool("cc")
            if cc then
                print("C Compiler (cc): " .. cc)
            else
                print("C Compiler (cc) not found")
            end
        else
            print("Toolchain for cc not found")
        end
    end)

打印出来是找不到cc
Toolchain for cc not found

**Comment by waruqi** [comment]
交叉编译通常不需要自定义 Toolchain，直接 `xmake f -p cross --sdk=/xxx` 就行了。文档有详细说明。https://xmake.io/#/zh-cn/guide/configuration?id=%e4%ba%a4%e5%8f%89%e7%bc%96%e8%af%91%e9%85%8d%e7%bd%ae

如果非要自定义 toolchain，一般也只需要设置 sdkdir 。。而你直接设置 cc 。。那么得先确保这些 编译器 已经存在于你的 PATH 中。

可以加 -vD 看下完整 logs 。。`xmake f -c -vD`

**Reply by waruqi** [reply]
另外， ld 应该设置成 g++ 而不是 ld

**Comment by liangzc1** [comment]
打印的问题，这样子不能正常获取到当前cc的

**Reply by waruqi** [reply]
我上面说的很明白了，你什么 logs 都不给，我怎么看，我没法凭空猜出来

> 如果非要自定义 toolchain，一般也只需要设置 sdkdir 。。而你直接设置 cc 。。那么得先确保这些 编译器 已经存在于你的 PATH 中。
> 可以加 -vD 看下完整 logs 。。xmake f -c -vD
